### PR TITLE
Issue #2245 - Intermittent "git ls-remote" request failures should not fail app reconciliation

### LINKED ISSUE #2245: Intermittent "git ls-remote" request failures should not fail app reconciliation
**Describe the bug**

Argo CD execute `git ls-remote` request during app reconciliation/creation/editing. If git request than the whole operation fails. Instead of failing the operation Argo CD should retry failed `git ls-remotes` before giving up.

**Version**

v1.2.0-rc2
**Logs**

```
 level=fatal msg="rpc error: code = InvalidArgument desc = application spec is invalid: InvalidSpecError: Unable to load app.yaml: rpc error: code = Unknown desc = Get <myrepourl>.git/info/refs?service=git-upload-pack: read tcp ip:port->ip:port: read: connection reset by peer"
```

### PR DESCRIPTION
**Pull request by alexmt**
Resolves #2245

### CODE REVIEW

File: util/git/client.go [hunk 1]
```diff
@@ -68,22 +64,28 @@ type nativeGitClient struct {
...
 }
 
-type factory struct{}
-
-func NewFactory() ClientFactory {
-	return &factory{}
-}
+func NewClient(rawRepoURL string, path string, creds Creds, insecure bool, enableLfs bool, reporter metrics.Reporter) (Client, error) {
+	retriesCount := 3
+	if retiesCountStr := os.Getenv(common.EnvGitRetriesCount); retiesCountStr != "" {
```
Comment (alexec) [inline util/git/client.go]: my not do in init()?

File: util/git/client.go [hunk 2]
```diff
@@ -309,7 +311,16 @@ func (m *nativeGitClient) Checkout(revision string) error {
 // Otherwise, it returns an error indicating that the revision could not be resolved. This method
 // runs with in-memory storage and is safe to run concurrently, or to be run without a git
 // repository locally cloned.
-func (m *nativeGitClient) LsRemote(revision string) (string, error) {
+func (m *nativeGitClient) LsRemote(revision string) (res string, err error) {
+	for attempt := 0; attempt < m.retriesCount; attempt++ {
+		if res, err = m.lsRemote(revision); err == nil {
```
Comment (alexec) [inline util/git/client.go]: thundering herd?

File: util/git/repo/repo.go [hunk 3]
```diff
@@ -68,7 +68,7 @@ func NewRepo(url string, creds git.Creds, insecure, enableLfs bool, disco func(r
 	if err != nil {
 		return nil, err
 	}
-	client, err := git.NewFactory().NewClient(url, workDir, creds, insecure, enableLfs, reporter)
+	client, err := git.NewClient(url, workDir, creds, insecure, enableLfs, reporter)
```
Comment (alexec) [inline util/git/repo/repo.go]: intentional change?

Comment (alexmt) [reply util/git/repo/repo.go]: Yes, looks like factory is a left-over after some refactoring. It is not injected anywhere and not used in tests anymore.

File: util/git/client.go (hunk 1 above)
Comment (alexmt) [reply util/git/client.go]: good idea, done

File: util/git/client.go (hunk 2 above)
Comment (alexmt) [reply util/git/client.go]: I guess you are right. It is not safe to enable it by default. Changed the default number of attempts to 1.

In the next PR I'm going to document all performance-related env variables and recommendations of how to use them

File: util/git/client.go [hunk 4]
```diff
@@ -23,6 +24,8 @@ import (
 	ssh2 "gopkg.in/src-d/go-git.v4/plumbing/transport/ssh"
 	"gopkg.in/src-d/go-git.v4/storage/memory"
 
+	"github.com/argoproj/argo-cd/common"
```
Comment (alexec) [inline util/git/client.go]: import weirdness

**Review by alexec — APPROVED** [review]

File: util/git/client.go (hunk 4 above)
Comment (alexmt) [reply util/git/client.go]: fixed

**Pull request merged**